Clearer HTTP error messages in `huggingface_hub`

Wauplin opened the pull request:
Fix https://github.com/huggingface/huggingface_hub/issues/989 (and duplicate https://github.com/huggingface/huggingface_hub/issues/989).

### What this PR do

The initial issue was about adding meaningful information to the HTTPError (especially advice like "create_commit expects the repo to exist" or "make sure you specific the right repo_id/repo_type"). This PR aims to solve that but also improve and harmonize how HTTP Errors are processed in `huggingface_hub`.

It introduces:
- **`hf_raise_for_status`** -> meant to be the single way to raise for status in HF Hub. Also meant to be reused "officially" by third-party libraries like `transformers` (at the moment `huggingface_hub.utils._errors._raise_for_status` is called).
- **`HfHubHTTPError`** -> a custom error that inherits from `response.HTTPError` and that is parents to the `RepositoryNotFoundError`, `RevisionNotFoundError`,... It handles the formatting of the Hub details in the error message.
- I also added [a small section in the documentation](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1019/en/package_reference/utilities#handling-http-errors) for `hf_raise_for_status` usage.

At the moment, `_raise_for_status`, `_raise_with_request_id` and `_raise_convert_bad_request` that were doing similar things are all kept as aliases for `hf_raise_for_status`. I deprecated their usage instead of removing them. I can also keep them without warning message but I like it less (cc @sgugger).

### Examples
I refined the error messages to have multi-line descriptions. I prefer it this way to make it clearer for the end-user but please let me know if you see a problem with that. Errors look like this:

```
huggingface_hub.utils._errors.RepositoryNotFoundError: 401 Client Error. (Request ID: PvMw_VjBMjVdMz53WKIzP)

Repository Not Found for url: https://huggingface.co/api/models/%3Cnon_existent_repository%3E.
Please make sure you specified the correct `repo_id` and `repo_type`.
If the repo is private, make sure you are authenticated.
Invalid username or password.
```

```
huggingface_hub.utils._errors.RepositoryNotFoundError: 404 Client Error. (Request ID: xTE4dk9e6-ZATpr3JeWgC)

Repository Not Found for url: https://hub-ci.huggingface.co/api/models/__DUMMY_TRANSFORMERS_USER__/repo_that_do_not_exist/preupload/main.
Please make sure you specified the correct `repo_id` and `repo_type`.
If the repo is private, make sure you are authenticated.
Note: Creating a commit assumes that the repo already exists on the Huggingface Hub. Please use `create_repo` if it's not the case.
```

```
huggingface_hub.utils._errors.RevisionNotFoundError: 404 Client Error. (Request ID: Mwhe_c3Kt650GcdKEFomX)

Revision Not Found for url: https://huggingface.co/bert-base-cased/resolve/%3Cnon-existent-revision%3E/config.json.
```

(cc @NielsRogge @merveenoyan @cakiki who showed interest in the issue)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1019). All of your documentation changes will be reflected on that endpoint.

sgugger commented:
The warnings will pollute the use of Transformers by any user using a source install of `huggingface_hub`, without them being able to do anything to silence them, so they should be removed.
This is not a normal deprecation cycle, just a temporary alias until the next release (as we discussed previously). The old names can all be removed in the last commit before releasing the new version of huggingface_hub.

Wauplin commented:
@sgugger I removed the warning from `_raise_for_status` that is used in Transformers in commit https://github.com/huggingface/huggingface_hub/pull/1019/commits/a545192ae06de3164bad70d8700ac75210856e51 and added a TODO there.

In general I'm not a big fan of having a temporary alias that we need to remember about when we do the release since it requires that we synchronize ourselves on the releases. So for this one I suggest that we keep the alias without warning and starting from just after v0.10 (so not in official v0.10 release) I will deprecate it/remove it considering that Transformers no longer uses it. Would that be ok ? The only difference is that we keep the alias for one more version and in between Transformers should be able to adapt but no warnings are thrown.

sgugger commented:
Yes, that works even better. Thanks!

sgugger commented on docs/source/package_reference/utilities.mdx:
```diff
@@ -49,14 +49,63 @@ Using these shouldn't be necessary if you use `huggingface_hub` and you don't mo
+
+### Raise for status
+
+`hf_raise_for_status` is meant to be the central method to "raise for status" from any
```
```suggestion
[`~utils.hf_raise_for_status`] is meant to be the central method to "raise for status" from any
```
to get the link to the doc.

sgugger commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1964,6 +1960,18 @@ def create_commit(
                 If the Hub API returns an HTTP 400 error (bad request)
             :class:`ValueError`:
                 If `create_pr` is `True` and revision is neither `None` nor `"main"`.
+            :class: `RepositoryNotFoundError`:
```
This is old sphinx syntax. Should probably be replaced by the new links? (same for the other values above)
```suggestion
            [`~utils.RepositoryNotFoundError`]:
```

sgugger commented on src/huggingface_hub/hf_api.py:
```diff
@@ -1964,6 +1960,18 @@ def create_commit(
+        `create_commit` assumes that the repo already exists on the Hub. If you get a
+        Client error 404, please make sure you are authenticated and that `repo_id` and
+        `repo_type` are set correctly. If repo does not exist, create it first using
+        [`~huggingface_hub.hf_api.create_repo`].
```
```suggestion
        [`~hf_api.create_repo`].
```
Nit: `huggingface_hub` is optional (but if you prefer to have it, feel free to ignore this).

sgugger commented on src/huggingface_hub/hf_api.py:
```diff
@@ -2123,6 +2139,15 @@ def upload_file(
+        `upload_file` assumes that the repo already exists on the Hub. If you get a
+        Client error 404, please make sure you are authenticated and that `repo_id` and
+        `repo_type` are set correctly. If repo does not exist, create it first using
+        [`~huggingface_hub.hf_api.create_repo`].
```
```suggestion
        [`~hf_api.create_repo`].
```

sgugger commented on src/huggingface_hub/hf_api.py:
```diff
@@ -2281,6 +2306,15 @@ def upload_folder(
+        `upload_folder` assumes that the repo already exists on the Hub. If you get a
+        Client error 404, please make sure you are authenticated and that `repo_id` and
+        `repo_type` are set correctly. If repo does not exist, create it first using
+        [`~huggingface_hub.hf_api.create_repo`].
```
```suggestion
        [`~hf_api.create_repo`].
```

sgugger commented on src/huggingface_hub/utils/_errors.py:
```diff
@@ -1,9 +1,67 @@
+    - Request id from "X-Request-Id" header if exists.
+    - Server error message if we can found one in the response body.
+
+    Usage:
```
Nit: if you call say `Example:` instead of `Usage:`, there will be a link to the example that can be shared.

Wauplin replied:
Good to know, thanks !

sgugger commented on src/huggingface_hub/utils/_errors.py:
```diff
@@ -71,131 +132,183 @@ class LocalEntryNotFoundError(EntryNotFoundError, FileNotFoundError, ValueError)
+    This helper is meant to be the unique method to raise_for_status when making a call
+    to the Hugging Face Hub.
 
+    Usage:
```
Same here.

sgugger commented on src/huggingface_hub/utils/_errors.py:
```diff
@@ -71,131 +132,183 @@ class LocalEntryNotFoundError(EntryNotFoundError, FileNotFoundError, ValueError)
+
+    Raises when the request has failed:
+
+        - [`~huggingface_hub.utils.RepositoryNotFoundError`]
+            If the repository to download from cannot be found. This may be because it
+            doesn't exist, because `repo_type` is not set correctly, or because the repo
+            is `private` and you do not have access.
+        - [`~huggingface_hub.utils.RevisionNotFoundError`]
+            If the repository exists but the revision couldn't be find.
+        - [`~huggingface_hub.utils.EntryNotFoundError`]
+            If the repository exists but the entry (e.g. the requested file) couldn't be
+            find.
+        - [`~huggingface_hub.utils.BadRequestError`]
+            If request failed with a HTTP 400 BadRequest error.
+        - [`~huggingface_hub.utils.HfHubHTTPError`]
+            If request failed for a reason not listed above.
```
```suggestion
        - [`~utils.RepositoryNotFoundError`]
            If the repository to download from cannot be found. This may be because it
            doesn't exist, because `repo_type` is not set correctly, or because the repo
            is `private` and you do not have access.
        - [`~utils.RevisionNotFoundError`]
            If the repository exists but the revision couldn't be find.
        - [`~utils.EntryNotFoundError`]
            If the repository exists but the entry (e.g. the requested file) couldn't be
            find.
        - [`~utils.BadRequestError`]
            If request failed with a HTTP 400 BadRequest error.
        - [`~utils.HfHubHTTPError`]
            If request failed for a reason not listed above.
```

sgugger reviewed (approved):
Left a couple of nits, but looks great otherwise!
Thanks!

LysandreJik commented on src/huggingface_hub/utils/_errors.py:
```diff
@@ -1,9 +1,67 @@
+            response=response,
+        )
+
+    def append_to_message(self, additional_message: str) -> None:
+        """Append additional information to the `HfHubHTTPError` initial message."""
+        self.args = (self.args[0] + additional_message,) + self.args[1:]
```
Very nice shortcut method

LysandreJik reviewed (approved):
This looks good to me! Thanks for working on it, @Wauplin, very happy to have a stable and maintained public API for `hf_raise_for_status`.
